python/ray/data/dataset_pipeline.py

```diff
@@ -179,11 +180,25 @@ def iter_batches(
         )
         self._stats.iter_total_s.add(time.perf_counter() - time_start)
 
-    def _iter_blocks(self) -> Iterator[Block]:
+    def _iter_blocks(self, prefetch_blocks: int = 0) -> Iterator[Block]:
         ds_wait_start = time.perf_counter()
+        latest_blocks = collections.deque(maxlen=prefetch_blocks + 1)
         for ds in self.iter_datasets():
             self._stats.iter_ds_wait_s.add(time.perf_counter() - ds_wait_start)
-            yield from ds._plan.execute().iter_blocks()
+            for block in ds._plan.execute().iter_blocks():
+                yield block
+                if len(latest_blocks) == latest_blocks.maxlen:
+                    block_ref = latest_blocks.popleft()
+                    # Eagerly clear the object before Python GC kicks in, which may 
+                    # have certain delay, to reduce the memory footprint.
+                    # Notes:
+                    # - We cannot make this optimization if it's eager dataset, as
+                    #   this will destruct the dataset.
+                    # - We can only clear the block if it has passed the prefetch
+                    #   window, i.e. the block is already consumed.
+                    if ds._lazy:
```

Comment:
Since dataset pipelines can only be consumed once, I think that clearing the freeing the blocks should always be allowed, even if the underlying Dataset is non-lazy.

Reply:
For repeat pipelines that wouldn't work though right?

Reply:
Ah yeah, I was thinking this in the context of block clearing after pipeline operations, but that's a good point. In that case, this lazy check should work, except for maybe the windowing case: https://github.com/ray-project/ray/blob/1f509ab331885ee4a43343d6fdb3a99875dbb813/python/ray/data/dataset.py#L3003

```python
ds = ray.data.read_parquet().map_batches()
pipe = ds.window()

for batch in pipe.iter_batches():
    # process batch, with the underlying blocks cleared.

for batch in ds.iter_batches():
    # this would fail
```

Reply:
Right, clearing non-lazy will result in error. Also dataset.iter_batches() is not destructive.

Reply:
> Also dataset.iter_batches() is not destructive.

Sure, but I think that the above example of windowing a dataset and consuming both the produced pipeline and the pre-windowed dataset will fail with the current "clear if lazy" semantics. Can you double-check that this works?

Reply:
The example works for me.